### PR TITLE
Mint

### LINKED ISSUE #98: We need minting
To allow users to deploy accounts, we need the minting of tokens to arbitrary addresses. Since users currently cannot deploy accounts, I label this as a bug.

Basically we need what we had here: https://0xspaceshard.github.io/starknet-devnet/docs/guide/mint-token/

### LINKED ISSUE #102: CLI parameters never used due to default StarknetConfig in Starknet::new
https://github.com/0xSpaceShard/starknet-devnet-rs/blob/83a4ba85fae8a08ab57a255f73df2059114128f4/crates/starknet/src/starknet/mod.rs#L119

### PR DESCRIPTION
**Pull request by FabijanC**
## Usage related changes

- Close #98
- Close #102
- Remove `lite` version of minting
  - It was first introduced to offer a faster version of minting, but now that we are in Rust, that's probably not so needed
  - Simplifies the logic compared to supporting both cases (lite and non-lite)

## Development related changes

- Introduce the chargeable account
  - Same what we had in Pythonic Devnet
  - Can be used when there is no other account to charge (e.g. when minting; also can be useful in tests)
- Minting is implemented as an invoke transaction signed by the chargeable account
  - Downsides of alternatively implementing as a direct state manipulation:
    - Involves uint256 arithmetic on our side
    - Prevents clear separation of block states (e.g. getStateUpdate would be wrong if calculated on block creation)
- Move some methods of trait `Accounted` to `Deployed` to better reflect their meaning
- Set ERC20 account properties
  - Actually only had to set the owner, otherwise minting couldn't be performed - only the owner can call that contract since it's Ownable
- Left one TODO comment (conversion `From<Felt> for BigUint`)
- Content of raw_execution.rs is mostly copied from starknet.rs
- Update starknet.rs crates to git tag `starknet-ff/0.3.4`

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes - not sure if the old docs should apply or if I should edit the README.md for this new/old minting feature
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet/src/starknet/predeployed.rs [hunk 1]
```diff
@@ -20,7 +28,26 @@ pub(crate) fn create_erc20() -> Result<SystemContract> {
...
+        ("ERC20_symbol", get_selector_from_name("ETH").unwrap().into()),
+        ("ERC20_decimals", 18.into()),
+        // necessary to set - otherwise minting txs cannot be performed
+        ("Ownable_owner", Felt::from_prefixed_hex_str(CHARGEABLE_ACCOUNT_ADDRESS).unwrap()),
+    ]
+    .into_iter()
+    .for_each(|(storage_var_name, storage_value)| {
+        let storage_var_address = get_storage_var_address(storage_var_name, &[]).unwrap();
+        let storage_key = ContractStorageKey::new(contract_address, storage_var_address);
+        state.change_storage(storage_key, storage_value).unwrap();
```
Comment (marioiordanov) [inline crates/starknet/src/starknet/predeployed.rs:46]: change_storage and get_storage_var_address could fail.

File: crates/starknet/src/starknet/mod.rs [hunk 2]
```diff
@@ -335,6 +353,55 @@ impl Starknet {
...
+    /// Creates an invoke tx for minting, using the chargeable account.
+    pub async fn mint(&mut self, address: ContractAddress, amount: u128) -> Result<Felt> {
+        let sufficiently_big_max_fee: u128 = self.config.gas_price as u128 * 1_000_000;
+        let chargeable_address_felt = Felt::from_prefixed_hex_str(CHARGEABLE_ACCOUNT_ADDRESS)?;
+        let nonce =
+            self.state.pending_state.get_nonce_at(&Address(chargeable_address_felt.into()))?;
+
+        let calldata = vec![
+            Felt::from(address).into(),
+            FieldElement::from_dec_str(&amount.to_string()).unwrap(), // `low` part of Uint256
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/starknet/mod.rs:366]: But I need the vector to contain `FieldElement` instances. For some reason, `FieldElement` implements `From` for u8, u16, u32 and u64, but not u128.

In any case, in the next line I can replace `FieldElement::from_dec_str("0").unwrap()` with `FieldElement::from(0u32)`

File: crates/starknet/src/starknet/predeployed.rs (hunk 1 above)
Comment (marioiordanov) [reply crates/starknet/src/starknet/predeployed.rs:46]: Btw If you are going to handle the result using ? sign, in the for_each method this cannot be done. You should use a simple for loop for that

**Review by marioiordanov — COMMENTED** [review]
comments

File: crates/starknet/src/starknet/mod.rs [hunk 3]
```diff
@@ -335,6 +353,55 @@ impl Starknet {
...
+            FieldElement::from_dec_str("0").unwrap(),                 // `high` part
+        ];
+
+        let erc20_address_felt = Felt::from_prefixed_hex_str(ERC20_CONTRACT_ADDRESS)?;
+        let raw_execution = RawExecution {
+            calls: vec![Call {
+                to: erc20_address_felt.into(),
+                selector: get_selector_from_name("mint").unwrap(),
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/starknet/mod.rs:374]: As long as our tests execute this line of the code, there is nothing to fear.

File: crates/starknet/src/starknet/predeployed.rs [hunk 4]
```diff
@@ -20,7 +28,26 @@ pub(crate) fn create_erc20() -> Result<SystemContract> {
     Ok(erc20_fee_contract)
 }
 
-pub(crate) fn create_udc20() -> Result<SystemContract> {
+/// Set initial values of ERC20 contract storage
+pub(crate) fn initialize_erc20(state: &mut StarknetState) {
+    let contract_address =
+        ContractAddress::new(Felt::from_prefixed_hex_str(ERC20_CONTRACT_ADDRESS).unwrap()).unwrap();
+    [
+        ("ERC20_name", get_selector_from_name("ether").unwrap().into()),
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/starknet/predeployed.rs:36]: Yes

File: crates/starknet/src/starknet/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/starknet/mod.rs]: Required a starknet-rs update. Addressed in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/101/commits/0a14458cb5a5d070a0123fbc7112bca1a3d70748

File: crates/starknet/src/starknet/mod.rs [hunk 5]
```diff
@@ -335,6 +353,55 @@ impl Starknet {
...
+                calldata: calldata.clone(),
+            }],
+            nonce: Felt::from(nonce.clone()).into(),
+            max_fee: FieldElement::from_dec_str(&sufficiently_big_max_fee.to_string()).unwrap(),
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/starknet/mod.rs]: Required a starknet-rs update. Addressed in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/101/commits/0a14458cb5a5d070a0123fbc7112bca1a3d70748

File: crates/types/src/felt.rs [hunk 6]
```diff
@@ -183,6 +185,13 @@ impl TryFrom<BigUint> for Felt {
     }
 }
 
+impl From<Felt> for BigUint {
+    fn from(felt: Felt) -> Self {
+        // TODO better use TryFrom without unwrapping, but how to map the errors in mint_token?
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/types/src/felt.rs]: Addressed in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/101/commits/a90abee8aa75c2bc8f113523b70060ab44e4b43b

**Pull request merged**